[Back+Front] Message Board

shakedhagag opened the pull request:
## MessageBoard
I have implemented the message board model:
1. Added a form that allows the **logged in user** to submit a message (if you're logged out you will be taken to login page)
2. added a url for the messageboard - mainly for testing right now
3. in Views I have added the render function of the message form and the messages from the db, this function handles the POST of the messages
4. added board.html 
5. altered 005 test data file for the migrations to work (msgID needed to be changed to msg)

Each course page will have this model of messages so that is a task for the upcoming days (@DeanCivlin is close to finish it).

closes #109

![image](https://user-images.githubusercontent.com/76105109/147388033-1d623eec-3d56-4b30-aa4d-b79600c3ec7b.png)

shakedhagag commented on AcadeMe/urls.py:
```diff
@@ -25,5 +25,7 @@
+    path('contact_us/', views.contact, name='contact_us'),
+    path('msgboard/', user_views.msgboard, name='msgboard')
+
 ]
```
The path to the msgboard link for testing

shakedhagag commented on AcadeMeData/forms.py:
```diff
@@ -32,3 +32,9 @@ def save(self, commit=True):
+class MessageForm(forms.ModelForm):
+    class Meta:
+        model = Messages
+        fields = ('text', 'userID')
```
The form of Messages

shakedhagag commented on AcadeMeData/migrations/0005_messageTags_test_data.py:
```diff
@@ -16,7 +16,7 @@ def generate_msgtagsdata(apps, schema_editor):
         with transaction.atomic():
             for id, msgID, userID in test_data:
-                msgTag = MessageTags(id=id, msgID=msgID, userID=userID)
+                msgTag = MessageTags(id=id, msg=msgID, userID=userID)
```
The little change that was needed for the migrations to work again

shakedhagag commented on AcadeMeData/views.py:
```diff
@@ -1,6 +1,8 @@
+from .forms import UserRegistrationForm, MessageForm
+from .models import Messages
+from django.contrib.auth.decorators import login_required
 
```
login required handles the auth of the user 
if youre logged out you will be redirected to login page

DeanCivlin reviewed (approved):
LGTM

DanWaldman reviewed (approved):
Looks good, in future PR we will be linking each board to specific course, nice work

razashtar93 reviewed (approved):
LGTM

Yarboa commented:
Nice work
Unfortunately there are missing tests. 
Please refer this https://github.com/redhat-beyond/AcadeMe/blob/main/contributing.md#guidelines-for-pr-reviews

shakedhagag commented:
> Nice work Unfortunately there are missing tests. Please refer this https://github.com/redhat-beyond/AcadeMe/blob/main/contributing.md#guidelines-for-pr-reviews

Done! added the test file for the feature
please review 🍔 
@Yarboa  🥲

Omeramsc commented on AcadeMe/urls.py:
```diff
@@ -25,5 +25,7 @@
-    path('contact_us/', views.contact, name='contact_us')
+    path('contact_us/', views.contact, name='contact_us'),
+    path('msgboard/', user_views.msgboard, name='msgboard')
+
```
no need for the empty line

shakedhagag replied:
got it will change

Omeramsc commented on AcadeMeData/tests/test_msgboard_feature.py:
```diff
@@ -0,0 +1,21 @@
+
+
+@pytest.fixture
+def user_example():
```
Any reason not to use `user_example` from `test_register` or from `test_user`? you should have all of those consolidated into a contest instead of rewriting them for each and every test module.

shakedhagag replied:
What do you mean? importing them or merging the tests in to one file?
If importing is the answer then I have tried and failed so I changed it to this way

Yarboa replied:
Please refer this review also 
https://github.com/redhat-beyond/AcadeMe/pull/103#discussion_r775504679

If you do not understand, please try friends in slack. understand it and propose a code reply

shakedhagag replied:
got it, fixed this in the recent push

Omeramsc reviewed (changes requested):
Left some comments you should address, 
BTW I appreciate the comments you added along the code, but if you feel like you need comments to explain what had been done on each stage, it is usually a sign the code is not readable enough or requires changes. 
I did however like the PR description, very helpful!

Omeramsc commented on AcadeMeData/tests/test_msgboard_feature.py:
```diff
@@ -0,0 +1,21 @@
+    response = client.get("/msgboard/")
+    assert response.status_code == 200
+    form = response.context["form"]
+    assert isinstance(form, MessageForm)
```
I'm not sure what this assertion is testing, can you explain?
anyway, you should check that a message inserted by a user appear in the messageboard.

shakedhagag replied:
Asserting if the form that is being filled in /msgboard/ page is an instance of MessageForm
which is the form that handles messages

BTW 
I have written comments on my code because past experience taught me that it would make the PR approved faster, I do believe it is well written

Omeramsc replied:
There's no use-case in which the form should not be a MessageForm, so I'm not sure why this assertion is needed. 
A better test would check the fields of the form we expect to see, or that it is saved correctly.

Yarboa reviewed (commented):
I think test coverage is not enough,
Please try to suggest more tests.

shakedhagag commented:
> I think test coverage is not enough, Please try to suggest more tests.

Added more tests and organized them accordingly
fixed confest and inserted more fixtures into it

Yarboa commented on AcadeMeData/tests/test_msgboard_feature.py:
```diff
@@ -0,0 +1,35 @@
+
+
+@pytest.mark.django_db
+def test_msgboard_feature(client, user_example, generate_msgboard):
```
I would add here validation of the following
      assert len(response.context['messages']) > 0
      and also search for the content

Yarboa replied:
sorry you are avoiding the reviews, again, check that you have returned messages, you should not check the form
Not instance, content

Yarboa replied:
nvm, at least test were added

Yarboa commented on AcadeMeData/tests/test_msgboard_feature.py:
```diff
@@ -0,0 +1,35 @@
+def test_msgboard_empty_message_when_page_loads(client, user_example):
+    client.force_login(user_example.user)
+    response = client.get("/msgboard/")
+    messages = list(get_messages(response.wsgi_request))
```
Consider this 
 len(response.context['messages'])  instead

shakedhagag replied:
The post test is giving me a hard time, checking everything from a to z, it only fills the form on /msgboard/ but not posting it and adding to the db
I think this is something that we can deal with later on, everything works manually, posting messages to msgboard is working perfectly 
(messages are being added to the DB) and other tests are fine too.
In order to get ahead with the project I would love for this pr to be merged and promise to deal with this specific test later on.

Omeramsc replied:
Tests are the most important part of the code, more than the functionality itself; it is not something we usually push aside for later. 
your code may work now, but a future feature (even if not related) may break it without you even noticing.

shakedhagag replied:
I understand and agree, but Its taking me to much time to solve it,
I have read the docs, countless stackoverflow questions
I have consulted mentors, friends, different teams and no one could help 🥲 
If you want I can show you what the problem is via zoom

Omeramsc replied:
I really do get your frustration. but it does not meet our definition of done. 
sometimes it is better not having a feature than having something that does not meet the standards we decided on (good coverage of tests in this situation). 
I am glad to hear it made you consult others, as this is a good way to learn. 

I would've liked to see the problem, but I don't think I'll be available for this before the demo...

shakedhagag replied:
I'll try again 🥲 
thx

Yarboa commented:
@shakedhagag look at test which is failing, you have no messages

shakedhagag commented:
> @shakedhagag look at test which is failing, you have no messages

Yeah I am still trying to figure out why POST is not working..
im on this since 4 pm ..

Yarboa commented:
It can not be reviewed, please resolve conflicts and push,

Yarboa commented:
I think this PR should wait for #116

shakedhagag commented:
rebased and pushed

Yarboa reviewed (commented):
Try to be more careful and review the comments
clients tests should check the responses empty and not empty
While forms could be tested under the views

it is useless to check instances, if data is not created due to a bug FYI

shakedhagag commented:
> Try to be more careful and review the comments clients tests should check the responses empty and not empty While forms could be tested under the views
> 
> it is useless to check instances, if data is not created due to a bug FYI

My bad I deleted it few days ago from the repo and after rebase it has been deleted
Im still working on it locally

razashtar93 commented on AcadeMe/views.py:
```diff
@@ -7,6 +7,7 @@ def app_layout(request):
 
 
 def homePage(request):
+    print(request.user)
```
do we need this print?

shakedhagag replied:
no hell we dont!
ill delete it now sorry mate

razashtar93 reviewed (approved):
Looks great!

Yarboa reviewed (approved):
LGTM, although i expected for more related to tests views against client experience
